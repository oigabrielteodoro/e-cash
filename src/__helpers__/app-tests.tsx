import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
