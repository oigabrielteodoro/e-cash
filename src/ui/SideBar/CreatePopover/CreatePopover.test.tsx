import React from 'react'

import { render, screen, userEvent, waitFor } from '__helpers__/app-tests'

import { CreatePopover } from '.'

describe('CreatePopover', () => {
  it('should be able render correctly', () => {
    render(<CreatePopover />)

    expect(screen.getByLabelText('create')).toBeInTheDocument()
  })

  it('should be able open popover when button is clicked', async () => {
    render(<CreatePopover />)

    const button = screen.getByLabelText('create')

    expect(button).toBeInTheDocument()

    userEvent.click(button)

    expect(await screen.findByLabelText('popover')).toBeInTheDocument()
    expect(await screen.findByLabelText('income')).toBeInTheDocument()
    expect(await screen.findByLabelText('expense')).toBeInTheDocument()
    expect(await screen.findByLabelText('credit card')).toBeInTheDocument()
    expect(await screen.findByLabelText('transaction')).toBeInTheDocument()
  })

  it('should be able open popover when button is clicked after close', async () => {
    render(<CreatePopover />)

    const button = screen.getByLabelText('create')

    expect(button).toBeInTheDocument()

    userEvent.click(button)

    expect(await screen.findByLabelText('popover')).toBeInTheDocument()
    expect(await screen.findByLabelText('income')).toBeInTheDocument()
    expect(await screen.findByLabelText('expense')).toBeInTheDocument()
    expect(await screen.findByLabelText('credit card')).toBeInTheDocument()
    expect(await screen.findByLabelText('transaction')).toBeInTheDocument()

    userEvent.click(button)

    await waitFor(() =>
      expect(screen.queryByLabelText('popover')).not.toBeInTheDocument(),
    )
  })
})
