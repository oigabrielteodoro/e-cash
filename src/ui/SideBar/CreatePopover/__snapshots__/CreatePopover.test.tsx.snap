// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePopover should be able match snapshot 1`] = `
.c7 {
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #ffffff;
  border-radius: 0.5rem;
  font-weight: 500;
  font-size: 1.6rem;
  -webkit-transition: 300ms;
  transition: 300ms;
  border: 0.1rem solid transparent;
  padding: 1.2rem 3.2rem;
  background: #3a67e1;
}

.c7:hover {
  opacity: 0.9;
  box-shadow: 0 0 0 0.4rem #a4b7ea30;
}

.c7:disabled {
  cursor: no-drop;
  color: #a1a7ae;
  background: #f5f5f5;
  border-color: #cfd7e3;
}

.c7:disabled:hover {
  box-shadow: 0 0 0 0.4rem #a1a7ae20;
}

.c8 {
  padding-left: 1.6rem;
  height: 4.2rem;
  margin-left: 0.4rem;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 0;
  width: 4.2rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  opacity: 0;
  visibility: hidden;
}

.c5 {
  position: absolute;
  color: #18171d;
  font-size: 1.2rem;
  font-weight: 500;
  background: #ffffff;
  padding: 1rem 1.6rem;
  border-radius: 0.5rem;
  white-space: nowrap;
  -webkit-transition: 300ms;
  transition: 300ms;
  border: 0.1rem solid #cfd7e3;
  text-align: center;
  z-index: 1;
  top: 50%;
  right: 0;
  -webkit-transform: translate(calc(100% + 1.6rem),-50%);
  -ms-transform: translate(calc(100% + 1.6rem),-50%);
  transform: translate(calc(100% + 1.6rem),-50%);
}

.c5::before {
  content: '';
  position: absolute;
  border-style: solid;
  border-width: 0.6rem;
  border-color: transparent;
}

.c6 {
  position: absolute;
  background: #ffffff;
  height: 1rem;
  width: 1rem;
  border: 0.1rem solid #cfd7e3;
  z-index: 1;
  top: 50%;
  left: calc(100% + 1.2rem);
  -webkit-transform: translateY(-50%) rotate(-45deg);
  -ms-transform: translateY(-50%) rotate(-45deg);
  transform: translateY(-50%) rotate(-45deg);
  border-bottom-color: transparent;
  border-right-color: transparent;
}

.c2 {
  position: relative;
}

.c2:hover .c3 {
  opacity: 1;
  visibility: visible;
}

.c0 {
  width: 100%;
  position: relative;
}

.c1 {
  width: 100%;
  border: 0;
  background: transparent;
  -webkit-transition: 300ms;
  transition: 300ms;
}

<div>
  <div
    class="c0"
    style="margin-top: 3rem; display: flex; align-items: center;"
  >
    <button
      aria-label="create"
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3 c4"
        >
          <span
            class="c5"
          >
            Create
          </span>
          <div
            class="c6"
          />
        </div>
        <div
          class="c7 c8"
        >
          <svg
            fill="currentColor"
            height="18"
            stroke="currentColor"
            stroke-width="0"
            t="1551322312294"
            version="1.1"
            viewBox="0 0 1024 1024"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs />
            <path
              d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
            />
            <path
              d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
            />
          </svg>
        </div>
      </div>
    </button>
  </div>
  <div
    class="Toastify"
  />
</div>
`;
